<a href="https://colab.research.google.com/github/Sowjenya/sentimental/blob/main/grammar_check.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip3 install pip==20.1.1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
!pip3 install git+https://github.com/PrithivirajDamodaran/Gramformer.git

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/PrithivirajDamodaran/Gramformer.git to /tmp/pip-req-build-lo94p3_l
  Running command git clone -q https://github.com/PrithivirajDamodaran/Gramformer.git /tmp/pip-req-build-lo94p3_l
     |████████████████████████████████| 4.2 MB 5.3 MB/s 
     |████████████████████████████████| 1.2 MB 42.9 MB/s 
     |████████████████████████████████| 50 kB 5.9 MB/s 
     |████████████████████████████████| 6.6 MB 36.1 MB/s 
     |████████████████████████████████| 140 kB 48.2 MB/s 
     |████████████████████████████████| 499 kB 47.5 MB/s 
     |████████████████████████████████| 86 kB 4.2 MB/s 
     |████████████████████████████████| 596 kB 39.5 MB/s 
     |████████████████████████████████| 1.8 MB 35.9 MB/s 
     |████████████████████████████████| 103 kB 64.6 MB/s 
  Created wheel for gramformer: filename=gramformer-1.0-py3-none-any.whl size=4462 sha256=ff487d510b0a176b6a9144d27b

In [4]:
import torch

In [6]:
from gramformer import Gramformer
import numpy as np
import pandas as pd

In [13]:
from google.colab import files


uploaded = files.upload()


Saving review_data.csv to review_data.csv


In [21]:
import pandas as pd
import io

df = pd.read_csv(io.BytesIO(uploaded['review_data.csv']))
print(df)
 

                                                    text  star  \
0                                        Anathi Khanyile     5   
1      Tony bahut funny hai Hill climbing racing my f...     5   
2                                                Teturwu     1   
3      Hoooooooooooyaaaaaaaaa what a game hoooooooooo...     5   
4                                      This game is nice     5   
...                                                  ...   ...   
29995                                 Plz my Vidos viral     5   
29996                                               Nice     5   
29997                  Disturbing too much as YouTube ad     1   
29998                                      Koub valo vai     5   
29999                          I want to grow my account     5   

                         app_id reviewDate  
0      com.fingersoft.hillclimb   18/03/21  
1      com.fingersoft.hillclimb   18/03/21  
2      com.fingersoft.hillclimb   18/03/21  
3      com.fingersoft.hillc

In [22]:
df.head()

text  star  \
0                                    Anathi Khanyile     5   
1  Tony bahut funny hai Hill climbing racing my f...     5   
2                                            Teturwu     1   
3  Hoooooooooooyaaaaaaaaa what a game hoooooooooo...     5   
4                                  This game is nice     5   

                     app_id reviewDate  
0  com.fingersoft.hillclimb   18/03/21  
1  com.fingersoft.hillclimb   18/03/21  
2  com.fingersoft.hillclimb   18/03/21  
3  com.fingersoft.hillclimb   18/03/21  
4  com.fingersoft.hillclimb   18/03/21

In [49]:
df

text
0                                        Anathi Khanyile
1      Tony bahut funny hai Hill climbing racing my f...
2                                                Teturwu
3      Hoooooooooooyaaaaaaaaa what a game hoooooooooo...
4                                      This game is nice
...                                                  ...
29995                                 Plz my Vidos viral
29996                                               Nice
29997                  Disturbing too much as YouTube ad
29998                                      Koub valo vai
29999                          I want to grow my account

[30000 rows x 1 columns]

In [66]:
s=str(df)

In [67]:
is_bad_rule = lambda rule: rule.message == 'Possible spelling mistake found.' and len(rule.replacements) and rule.replacements[0][0].isupper()

In [68]:
pip install language-tool-python

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [69]:
import language_tool_python
tool = language_tool_python.LanguageTool('en-US') 

In [58]:
matches = tool.check(s)

In [70]:
matches = [rule for rule in matches if not is_bad_rule(rule)]

In [71]:
data=language_tool_python.utils.correct(s, matches)

In [72]:
data

'                                                    text\n0                                        Anathi Khanyile\n1      Tony bahut funny hai Hill climbing racing my f...\n2                                                Teturwu\n3      Hoooooooooooyaaaaaaaaa what a game hoooooooooo...\n4                                      This game is nice\n...                                                  ...\n29995                                 Plz my Vidos viral\n29996                                               Nice\n29997                  Disturbing too much as YouTube ad\n29998                                      Koub valo vai\n29999                          I want to grow my account\n\n[30000 rows x 1 columns]'

                                                    text
0                                        Anathi Khanyile
1      Tony bahut funny hai Hill climbing racing my f...
2                                                Teturwu
3      Hoooooooooooyaaaaaaaaa what a game hoooooooooo...
4                                      This game is nice
...                                                  ...
29995                                 Plz my Vidos viral
29996                                               Nice
29997                  Disturbing too much as YouTube ad
29998                                      Koub valo vai
29999                          I want to grow my account

[30000 rows x 1 columns]

I gave done this correction using language tool python -- https://pypi.org/project/language-tool-python/.

In [74]:
! pip install -q pyspark==3.1.2 spark-nlp

     |████████████████████████████████| 212.4 MB 56 kB/s 
     |████████████████████████████████| 145 kB 42.9 MB/s 
     |████████████████████████████████| 198 kB 10.2 MB/s 


In [77]:
import pandas as pd
import numpy as np
import json
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
from pyspark.ml import Pipeline
from pyspark.sql import SparkSession
import pyspark.sql.functions as F
from sparknlp.annotator import *
from sparknlp.base import *
import sparknlp
from sparknlp.pretrained import PretrainedPipeline

In [78]:
spark = sparknlp.start()

In [79]:
document_assembler = DocumentAssembler()\
  .setInputCol("text")\
  .setOutputCol("document")

tokenizer = RecursiveTokenizer()\
  .setInputCols(["document"])\
  .setOutputCol("token")\
  .setPrefixes(["\"", "(", "[", "\n"])\
  .setSuffixes([".", ",", "?", ")","!", "‘s"])

spell_model = ContextSpellCheckerModel\
    .pretrained('spellcheck_dl')\
    .setInputCols("token")\
    .setOutputCol("corrected")

finisher = Finisher().setInputCols("corrected")

light_pipeline = Pipeline(stages = [
                                    document_assembler,
                                    tokenizer,
                                    spell_model,
                                    finisher
                                    ])
## For comparison
full_pipeline = Pipeline(
    stages = [
              document_assembler,
              tokenizer,
              spell_model
  ])

empty_ds = spark.createDataFrame([[""]]).toDF("text")
pipeline_model = full_pipeline.fit(empty_ds)
l_pipeline_model = LightPipeline(light_pipeline.fit(empty_ds))

spellcheck_dl download started this may take some time.
Approximate size to download 95.1 MB
[OK!]


In [80]:
input_list = [data]

In [81]:
df = spark.createDataFrame(pd.DataFrame({"text": input_list}))
result = pipeline_model.transform(df)

In [82]:
light_result = l_pipeline_model.annotate(input_list[0])

In [84]:
exploded = F.explode(F.arrays_zip('token.result', 'corrected.result'))
select_expression_0 = F.expr("cols['0']").alias("original")
select_expression_1 = F.expr("cols['1']").alias("corrected")
result.select(exploded.alias("cols")) \
    .select(select_expression_0, select_expression_1).show(truncate=False)

+--------+---------+
|original|corrected|
+--------+---------+
|text    |text     |
|
       |
        |
|0       |0        |
|Anathi  |Anthe    |
|Khanyile|Handle   |
|
       |
        |
|1       |1        |
|Tony    |Tony     |
|bahut   |but      |
|funny   |funny    |
|hai     |hai      |
|Hill    |Hill     |
|climbing|climbing |
|racing  |racing   |
|my      |my       |
|f       |f        |
|.       |.        |
|.       |.        |
|.       |.        |
|
       |
        |
+--------+---------+
only showing top 20 rows

